compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2577,7 +2584,12 @@ public Boolean visit(BTupleType t, BType s) {
                     return false;
                 }
             }
-            return true;
+
+            if (sourceRestType == null || targetRestType == symTable.noType) {
+                return true;
```

Comment:
Do we need null check here since the logic in line 2575. If we need to allow **noType** for `targetRestType ` we need to change previous logic right? Now when we reach here both `sourceRestType` and `targetRestType` are null.

```Java
if ((sourceRestType == null || targetRestType == null) && sourceRestType != targetRestType) {
   return false;
}
```

Reply:
No, right? 

We can reach here if 
- `sourceRestType` and `targetRestType` are both null or
- when neither is null 

So `sourceRestType == null || targetRestType == symTable.noType` will evaluate to true if 
- both `sourceRestType` and `targetRestType` are null or
- `sourceRestType` is not null and `targetRestType` is `symTable.noType`?

Reply:
Is the output for this situation as False correct?
`sourceRestType == null && targetRestType == symTable.noType`

Reply:
Yes, this noType case is for scenarios where var is used. For example,

```
var [a, ...b] = getTuple();  

function getTuple() returns [int] => [1];
```

The same type check is done with a tuple created for the binding pattern used with var - `[noType, noType...]`. Here `targetRestType` is `symTable.noType` and `sourceRestType` is null, so this returning false and causing a compilation error is correct.

Reply:
Okay I agree but my concern is that logic is complex here.
1. When `sourceRestType == null` we are not actually testing whether it is null but whether both are null. 
2. ```sourceRestType == null || targetRestType == symTable.noType```
  here looking at this line I would expect `sourceRestType == null && targetRestType == symTable.noType` to be true too but it false.

Reply:
Agreed that you need to have the preceding logic in mind. But was trying to avoid duplicate checks. 

Also, the noType scenario is kind of a special case.